nx/lib/nx.ex

```diff
@@ -3551,6 +3551,165 @@ defmodule Nx do
     end)
   end
 
+  @doc """
+  Transforms the leading dimension in a tensor into a vectorized axis.
+
+  Nested vectorizations will "remove" the leading axis from the shape
+  and append it to the vectorized axes
+
+  {2, 3, 4}
+  |> vectorize(:first) # => {3, 4} vec [first: 2]
+  |> vectorize(:second) # => {4} vec [first: 2, second: 3]
+  """
+  @doc type: :shape
+  def vectorize(tensor, name)
+
+  def vectorize(_tensor, name) when not is_atom(name) do
+    raise ArgumentError, "name for new vectorized axis must be an atom, got: #{inspect(name)}"
+  end
+
+  def vectorize(%Nx.Tensor{shape: {}}, _name) do
+    raise ArgumentError, "cannot vectorize tensor of rank 0"
+  end
+
+  def vectorize(%Nx.Tensor{names: names, shape: shape, vectorized_axes: vec_axes} = tensor, name) do
+    size = elem(shape, 0)
+    new_shape = Tuple.delete_at(shape, 0)
+    names = Enum.drop(names, 1)
```

Comment:
```suggestion
    names = tl(names)
```

Reply:
tl raises if the list is empty

Reply:
Oh, but I guess it's guaranteed that it's non-empty here